### PR TITLE
Add support for elfCuster/elfMachine force delete

### PR DESCRIPTION
**Pull request by Levi080513**

### CODE REVIEW

File: api/v1beta1/elfcluster_types.go [hunk 1]
```diff
@@ -26,6 +26,10 @@ const (
 	// resources associated with ElfCluster before removing it from the
 	// API server.
 	ClusterFinalizer = "elfcluster.infrastructure.cluster.x-k8s.io"
+
+	// ElfClusterForceDeleteAnnotation is an annotation that can be applied to ElfCluster
+	// skipping infra resource deletion when delete ElfCluster
+	ElfClusterForceDeleteAnnotation = "kubesmart.smtx.io/force-delete-cluster"
```
Comment (jessehu) [inline api/v1beta1/elfcluster_types.go:32]: 这里不需要有kubesmart字样, 可使用 `cape.infrastructure.cluster.x-k8s.io/force-delete-cluster` 
@huaqing1994

File: api/v1beta1/elfcluster_types.go [hunk 2]
```diff
@@ -84,6 +88,14 @@ func (c *ElfCluster) GetTower() Tower {
 	return c.Spec.Tower
 }
 
+func (c *ElfCluster) ForceDelete() bool {
```
Comment (jessehu) [inline api/v1beta1/elfcluster_types.go:91]: => hasForceDeleteCluster

File: controllers/elfcluster_controller.go [hunk 3]
```diff
@@ -207,6 +207,14 @@ func (r *ElfClusterReconciler) reconcileDelete(ctx *context.ClusterContext) (rec
 		return reconcile.Result{RequeueAfter: config.DefaultRequeue}, nil
 	}
 
+	// if cluster need to force delete, skipping infra resource deletion and remove the finalizer.
+	if ctx.ElfCluster.ForceDelete() {
+		ctx.Logger.Info("skipping infra resource deletion")
+
+		ctrlutil.RemoveFinalizer(ctx.ElfCluster, infrav1.ClusterFinalizer)
```
Comment (jessehu) [inline controllers/elfcluster_controller.go:214]: 只是跳过 调用Tower API 去删除 ElfMachine对应的 VM， 并不是跳过 ElfCluster对象。

File: controllers/elfmachine_controller.go [hunk 4]
```diff
@@ -307,6 +307,14 @@ func (r *ElfMachineReconciler) reconcileDelete(ctx *context.MachineContext) (rec
 
 	conditions.MarkFalse(ctx.ElfMachine, infrav1.VMProvisionedCondition, clusterv1.DeletingReason, clusterv1.ConditionSeverityInfo, "")
 
+	// if cluster need to force delete, skipping VM deletion and remove the finalizer.
+	if ctx.ElfCluster.ForceDelete() {
+		ctx.Logger.Info("skipping VM deletion")
```
Comment (jessehu) [inline controllers/elfmachine_controller.go:312]: ctx.Logger.Info("Skip VM deletion due to the force-delete-cluster annotation")

**Review by jessehu — APPROVED** [review]

File: controllers/elfcluster_controller.go (hunk 3 above)
Comment (Levi080513) [reply controllers/elfcluster_controller.go:214]: elfCluster 删除时会去调用Tower API清理label，如果这部分不跳过的话，删除elfCluster会一直卡住

Comment (jessehu) [reply controllers/elfcluster_controller.go:214]: 那就在清理label那个地方跳过

File: api/v1beta1/elfcluster_types.go [hunk 5]
```diff
@@ -26,6 +26,10 @@ const (
 	// resources associated with ElfCluster before removing it from the
 	// API server.
 	ClusterFinalizer = "elfcluster.infrastructure.cluster.x-k8s.io"
+
+	// ElfClusterForceDeleteAnnotation is an annotation that can be applied to ElfCluster
```
Comment (jessehu) [inline api/v1beta1/elfcluster_types.go:30]: // ElfClusterForceDeleteAnnotation means to skip the deletion of infrastructure resources in Tower (e.g. VM and labels)
// when deleting an ElfCluster. This is useful when the Tower server or SMTX ELF cluster is disconnected.

**Pull request merged**